spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/ChangelogIterator.java

```diff
@@ -0,0 +1,162 @@
+ * copy-on-write(COW) mechanism. For example, there are row1 (id=1, data='a') and row2 (id=2,
+ * data='b') in a data file, if we only delete row2, the COW will copy row1 to a new data file and
+ * delete the whole old data file. The changelog table will have two delete-rows(row1 and row2), and
+ * one insert-row(row1). Row1 is a carry-over row.
+ *
+ * <p>The iterator marks the delete-row and insert-row to be the update-rows. For example, these two
+ * rows
+ *
+ * <ul>
+ *   <li>(id=1, data='a', op='DELETE')
+ *   <li>(id=1, data='b', op='INSERT')
+ * </ul>
+ *
+ * will be marked as update-rows:
+ *
+ * <ul>
+ *   <li>(id=1, data='a', op='UPDATE_BEFORE')
+ *   <li>(id=1, data='b', op='UPDATE_AFTER')
+ * </ul>
+ */
+public class ChangelogIterator implements Iterator<Row>, Serializable {
+  private static final String DELETE = ChangelogOperation.DELETE.name();
+  private static final String INSERT = ChangelogOperation.INSERT.name();
+  private static final String UPDATE_BEFORE = ChangelogOperation.UPDATE_BEFORE.name();
+  private static final String UPDATE_AFTER = ChangelogOperation.UPDATE_AFTER.name();
+
+  private final Iterator<Row> rowIterator;
+  private final int changeTypeIndex;
+  private final List<Integer> partitionIdx;
+
+  private Row cachedRow = null;
+
+  public ChangelogIterator(
+      Iterator<Row> rowIterator, int changeTypeIndex, List<Integer> partitionIdx) {
+    this.rowIterator = rowIterator;
+    this.changeTypeIndex = changeTypeIndex;
+    this.partitionIdx = partitionIdx;
+  }
+
+  @Override
+  public boolean hasNext() {
+    if (cachedRow != null) {
+      return true;
+    }
+    return rowIterator.hasNext();
+  }
+
+  @Override
+  public Row next() {
+    // if there is an updated cached row, return it directly
+    if (updated(cachedRow)) {
+      Row row = cachedRow;
+      cachedRow = null;
+      return row;
+    }
+
+    Row currentRow = currentRow();
+
+    if (rowIterator.hasNext()) {
+      GenericRowWithSchema nextRow = (GenericRowWithSchema) rowIterator.next();
+      cachedRow = nextRow;
+
+      if (updateOrCarryoverRecord(currentRow, nextRow)) {
+        Row[] rows = update((GenericRowWithSchema) currentRow, nextRow);
+
+        currentRow = rows[0];
+        cachedRow = rows[1];
+      }
+    }
+
+    return currentRow;
+  }
+
+  private Row[] update(GenericRowWithSchema currentRow, GenericRowWithSchema nextRow) {
+    GenericInternalRow deletedRow = new GenericInternalRow(currentRow.values());
+    GenericInternalRow insertedRow = new GenericInternalRow(nextRow.values());
+
+    if (isCarryoverRecord(deletedRow, insertedRow)) {
+      // set carry-over rows to null for filtering out later
+      return new Row[] {null, null};
+    } else {
+      deletedRow.update(changeTypeIndex, UPDATE_BEFORE);
+      insertedRow.update(changeTypeIndex, UPDATE_AFTER);
+
+      return new Row[] {
+        RowFactory.create(deletedRow.values()), RowFactory.create(insertedRow.values())
```

Comment:
Why are we making new rows here? Can we not just use our GenericInternalRows here?

Reply:
`GenericInternalRow` is not a subclass of `Row`, we need to return a `Row` instance here.